lib/code_corps/stripe_service/webhook_processing/webhook_processor.ex

```diff
@@ -5,79 +5,104 @@ defmodule CodeCorps.StripeService.WebhookProcessing.WebhookProcessor do
+  the handler module as the second parameter.
 
   ## Returns
 
-  * `{:ok, :ignored_by_environment}` if the event was ignored due to environment mismatch
-  * `{:ok, :enqueued}` if the event will be handled
+  - `{:ok, pid}` if the event will be handled
+  - `{:error, :ignored_by_environment}` if the event was ignored due to
+    environment mismatch
 
   ## Note
 
   Stripe events can have their `livemode` property set to `true` or `false`.
-  A livemode event should be handled by the production environment, while all other environments
-  handle non-livemode events.
+  A livemode `true` event should be handled by the production environment,
+  while all other environments handle livemode `false` events.
   """
-  def process_async(%{} = json, handler) do
-    case event_matches_environment?(json) do
-      true -> do_process_async(json, handler)
-      false -> {:ok, :ignored_by_environment}
+  def process_async(%{"id" => id, "livemode" => livemode, "user_id" => user_id} = json, handler) do
+    case event_matches_environment?(livemode) do
+      true -> do_process_async(id, user_id, handler, json)
+      false -> {:error, :ignored_by_environment}
     end
   end
-
-  defp do_process_async(json, handler) do
-    Task.Supervisor.start_child(:webhook_processor, fn -> do_process(json, handler) end)
+  def process_async(%{"id" => id, "livemode" => livemode} = json, handler) do
+    case event_matches_environment?(livemode) do
+      true -> do_process_async(id, nil, handler, json)
+      false -> {:error, :ignored_by_environment}
+    end
   end
 
-  defp event_matches_environment?(%{"livemode" => livemode}) do
-    case Application.get_env(:code_corps, :stripe_env) do
-      :prod -> livemode
-      _ -> !livemode
-    end
+  defp do_process_async(id, user_id, handler, json) do
+    Task.Supervisor.start_child(:webhook_processor, fn -> do_process(id, user_id, handler, json) end)
   end
 
-  defp do_process(%{"id" => event_id, "type" => event_type} = json, handler) do
-    with {:ok, %StripeEvent{} = event} <- find_or_create_event(event_id, event_type) do
-      case handler.handle_event(json) |> Tuple.to_list do
-        [:ok, :unhandled_event] -> event |> set_unhandled
-        [:ok | _results]        -> event |> set_processed
-        [:error | _error]       -> event |> set_errored
-      end
+  defp do_process(id, user_id, handler, json) do
+    with {:ok, %Stripe.Event{id: api_event_id, type: api_event_type, user_id: api_user_id}} <- retrieve_event_from_api(id, user_id),
+         {:ok, endpoint} <- infer_endpoint_from_handler(handler),
+         {:ok, %StripeEvent{} = event} <- find_or_create_event(api_event_id, api_event_type, api_user_id, endpoint)
+    do
+      handle_event(json, event, handler)
     else
       {:error, :already_processing} -> nil
     end
   end
 
-  defp find_or_create_event(id_from_stripe, type) do
+  defp event_matches_environment?(livemode) do
+    case Application.get_env(:code_corps, :stripe_env) do
```

Comment:
I tried thinking about how to improve this because it reads strange, but think it's isolated enough to not make a difference, and is kind of addressed in the moduledoc.